File: trusera-sdk-js/src/interceptor.ts
import type { TruseraClient } from "./client.js";
import { EventType, createEvent } from "./events.js";

/**
 * Enforcement modes for policy violations.
 */
export type EnforcementMode = "block" | "warn" | "log";

/**
 * Configuration for the HTTP interceptor.
 */
export interface InterceptorOptions {
  /** How to handle policy violations (default: "log") */
  enforcement?: EnforcementMode;
  /** Cedar policy service URL for runtime policy checks */
  policyUrl?: string;
  /** URL patterns to exclude from interception (regex strings) */
  excludePatterns?: string[];
  /** Enable debug logging */
  debug?: boolean;
}

/**
 * Policy evaluation request sent to Cedar policy service.
 */
interface PolicyEvaluationRequest {
  principal: string;
  action: string;
  resource: string;
  context: Record<string, unknown>;
}

/**
 * Policy evaluation response from Cedar service.
 */
interface PolicyEvaluationResponse {
  decision: "Allow" | "Deny";
  reasons?: string[];
}

/**
 * Global interceptor singleton state.
 * Only one interceptor can be active at a time.
 */
let activeInterceptor: TruseraInterceptor | null = null;
let originalFetch: typeof globalThis.fetch | null = null;

/**
 * HTTP interceptor for AI agent outbound traffic.
 * Monkey-patches globalThis.fetch to intercept all HTTP calls,
 * evaluate them against Cedar policies, and track events.
 *
 * This is the core differentiator of the Trusera SDK - transparent
 * runtime monitoring without code changes.
 *
 * @example
 * ```typescript
 * const client = new TruseraClient({ apiKey: "tsk_xxx" });
 * const interceptor = new TruseraInterceptor();
 *
 * interceptor.install(client, {
 *   enforcement: "block",
 *   policyUrl: "https://policy.trusera.io/evaluate",
 *   excludePatterns: ["^https://api\\.trusera\\.io/.*"]
 * });
 *
 * // All fetch calls are now intercepted
 * await fetch("https://api.github.com/repos/test"); // Tracked + policy-checked
 *
 * interceptor.uninstall(); // Restore original fetch
 * ```
 */
export class TruseraInterceptor {
  private client: TruseraClient | null = null;
  private options: Required<InterceptorOptions> = {
    enforcement: "log",
    policyUrl: "",
    excludePatterns: [],
    debug: false,
  };
  private excludeRegexes: RegExp[] = [];
  private isInstalled = false;

  /**
   * Installs the HTTP interceptor by patching globalThis.fetch.
   * Only one interceptor can be active at a time.
   *
   * @param client - TruseraClient instance for event tracking
   * @param options - Interceptor configuration
   * @throws Error if another interceptor is already installed
   */
  install(client: TruseraClient, options: InterceptorOptions = {}): void {
    if (activeInterceptor !== null && activeInterceptor !== this) {
      throw new Error("Another TruseraInterceptor is already installed. Call uninstall() first.");
    }

    if (this.isInstalled) {
      this.log("Interceptor already installed");
      return;
    }

    this.client = client;
    this.options = {
      enforcement: options.enforcement ?? "log",
      policyUrl: options.policyUrl ?? "",
      excludePatterns: options.excludePatterns ?? [],
      debug: options.debug ?? false,
    };

    // Compile exclude patterns to regexes
    this.excludeRegexes = this.options.excludePatterns.map((pattern) => new RegExp(pattern));

    // Save original fetch and install interceptor
    if (originalFetch === null) {
      originalFetch = globalThis.fetch;
    }

    globalThis.fetch = this.createInterceptedFetch();
    activeInterceptor = this;
    this.isInstalled = true;

    this.log("Interceptor installed", {
      enforcement: this.options.enforcement,
      policyUrl: this.options.policyUrl,
      excludePatterns: this.options.excludePatterns.length,
    });
  }

  /**
   * Uninstalls the interceptor and restores original fetch.
   */
  uninstall(): void {
    if (!this.isInstalled) {
      return;
    }

    if (originalFetch !== null) {
      globalThis.fetch = originalFetch;
      originalFetch = null; // Reset for next install
    }

    activeInterceptor = null;
    this.isInstalled = false;
    this.client = null;

    this.log("Interceptor uninstalled");
  }

  /**
   * Creates the intercepted fetch function.
   * This is the core of the interception logic.
   */
  private createInterceptedFetch(): typeof globalThis.fetch {
    const self = this;

    return async function interceptedFetch(
      input: RequestInfo | URL,
      init?: RequestInit
    ): Promise<Response> {
      // Ensure we have original fetch to fall back to
      if (originalFetch === null) {
        throw new Error("Original fetch not available");
      }

      const url = typeof input === "string" ? input : input instanceof URL ? input.href : input.url;
      const method = init?.method ?? "GET";

      // Check if URL should be excluded
      if (self.shouldExclude(url)) {
        self.log("Skipping excluded URL", { url });
        return originalFetch(input, init);
      }

      self.log("Intercepting request", { method, url });

      // Extract request details
      const requestData = await self.extractRequestData(input, init);

      // Create tracking event
      const event = createEvent(
        EventType.API_CALL,
        self.generateEventName(url, method),
        {
          method,
          url,
          headers: requestData.headers,
          body: requestData.body,
        },
        {
          interception_mode: self.options.enforcement,
        }
      );

      // Evaluate policy if configured
      if (self.options.policyUrl) {
        const policyDecision = await self.evaluatePolicy(url, method, requestData);

        if (policyDecision.decision === "Deny") {
          self.log("Policy violation detected", { url, reasons: policyDecision.reasons });

          // Track violation event
          const violationEvent = createEvent(
            EventType.API_CALL,
            `${self.generateEventName(url, method)}.policy_violation`,
            {
              ...event.payload,
              policy_decision: "Deny",
              policy_reasons: policyDecision.reasons,
            }
          );

          self.client?.track(violationEvent);

          // Handle based on enforcement mode
          if (self.options.enforcement === "block") {
            throw new Error(
              `[Trusera] Policy violation: ${policyDecision.reasons?.join(", ") ?? "Request denied"}`
            );
          } else if (self.options.enforcement === "warn") {
            console.warn(
              `[Trusera] Policy violation (allowed): ${policyDecision.reasons?.join(", ") ?? "Request denied"}`
            );
          }
        } else {
          self.log("Policy check passed", { url });
        }
      }

      // Track the API call event
      self.client?.track(event);

      // Execute the actual request
      const startTime = Date.now();
      let response: Response;
      let error: Error | undefined;

      try {
        response = await originalFetch(input, init);
      } catch (err) {
        error = err as Error;
        self.log("Request failed", { url, error: error.message });

        // Track error event
        const errorEvent = createEvent(
          EventType.API_CALL,
          `${self.generateEventName(url, method)}.error`,
          {
            ...event.payload,
            error: error.message,
            duration_ms: Date.now() - startTime,
          }
        );
        self.client?.track(errorEvent);

        throw error;
      }

      // Track response
      const responseEvent = createEvent(
        EventType.API_CALL,
        `${self.generateEventName(url, method)}.response`,
        {
          ...event.payload,
          status: response.status,
          status_text: response.statusText,
          duration_ms: Date.now() - startTime,
          response_headers: (() => { const h: Record<string, string> = {}; response.headers.forEach((v, k) => { h[k] = v; }); return h; })(),
        }
      );
      self.client?.track(responseEvent);

      return response;
    };
  }

  /**
   * Checks if a URL should be excluded from interception.
   */
  private shouldExclude(url: string): boolean {
    return this.excludeRegexes.some((regex) => regex.test(url));
  }

  /**
   * Extracts request data for policy evaluation and tracking.
   */
  private async extractRequestData(
    _input: RequestInfo | URL,
    init?: RequestInit
  ): Promise<{ headers: Record<string, string>; body: string | null }> {
    const headers: Record<string, string> = {};

    // Extract headers
    if (init?.headers) {
      if (init.headers instanceof Headers) {
        init.headers.forEach((value, key) => {
          headers[key] = value;
        });
      } else if (Array.isArray(init.headers)) {
        for (const [key, value] of init.headers) {
          headers[key] = value;
        }
      } else {
        Object.assign(headers, init.headers);
      }
    }

    // Extract body (if present and serializable)
    let body: string | null = null;
    if (init?.body) {
      if (typeof init.body === "string") {
        body = init.body;
      } else if (init.body instanceof URLSearchParams) {
        body = init.body.toString();
      } else if (init.body instanceof FormData) {
        body = "[FormData]";
      } else {
        body = "[Binary data]";
      }
    }

    return { headers, body };
  }

  /**
   * Evaluates request against Cedar policies.
   */
  private async evaluatePolicy(
    url: string,
    method: string,
    requestData: { headers: Record<string, string>; body: string | null }
  ): Promise<PolicyEvaluationResponse> {
    if (!this.options.policyUrl || !originalFetch) {
      return { decision: "Allow" };
    }

    try {
      const evaluationRequest: PolicyEvaluationRequest = {
        principal: this.client?.getAgentId() ?? "unknown-agent",
        action: `http.${method.toLowerCase()}`,
        resource: url,
        context: {
          headers: requestData.headers,
          body_present: requestData.body !== null,
          timestamp: new Date().toISOString(),
        },
      };

      const response = await originalFetch(this.options.policyUrl, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(evaluationRequest),
      });

      if (!response.ok) {
        console.error(`[Trusera] Policy evaluation failed: ${response.status}`);
        return { decision: "Allow" }; // Fail open
      }

      return (await response.json()) as PolicyEvaluationResponse;
    } catch (error) {
      console.error("[Trusera] Policy evaluation error:", error);
      return { decision: "Allow" }; // Fail open on error
    }
  }

  /**
   * Generates a structured event name from URL and method.
   */
  private generateEventName(url: string, method: string): string {
    try {
      const urlObj = new URL(url);
      const hostname = urlObj.hostname.replace(/\./g, "_");
      const path = urlObj.pathname.split("/").filter(Boolean).join("_") || "root";
      return `http.${method.toLowerCase()}.${hostname}.${path}`;
    } catch {
      return `http.${method.toLowerCase()}.invalid_url`;
    }
  }

  private log(message: string, data?: Record<string, unknown>): void {
    if (this.options.debug) {
      console.log(`[Trusera Interceptor] ${message}`, data ?? "");
    }
  }
}

File: trusera-sdk-go/interceptor.go
package trusera

import (
	"bytes"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strings"
)

const maxBodySnippet = 500

// EnforcementMode determines how policy violations are handled
type EnforcementMode string

const (
	ModeLog   EnforcementMode = "log"   // Record but allow all requests
	ModeWarn  EnforcementMode = "warn"  // Log warnings for blocked patterns but allow
	ModeBlock EnforcementMode = "block" // Reject blocked requests with error
)

// InterceptorOptions configures the HTTP interceptor
type InterceptorOptions struct {
	Enforcement     EnforcementMode
	ExcludePatterns []string // URL patterns to skip interception
	BlockPatterns   []string // URL patterns to block (for testing enforcement)
}

// WrapHTTPClient wraps an http.Client to intercept all outbound requests
func WrapHTTPClient(client *http.Client, truseraClient *Client, opts InterceptorOptions) *http.Client {
	if client == nil {
		client = &http.Client{}
	}

	transport := client.Transport
	if transport == nil {
		transport = http.DefaultTransport
	}

	client.Transport = &interceptingTransport{
		base:   transport,
		client: truseraClient,
		opts:   opts,
	}

	return client
}

// interceptingTransport wraps http.RoundTripper
type interceptingTransport struct {
	base   http.RoundTripper
	client *Client
	opts   InterceptorOptions
}

// RoundTrip intercepts and records HTTP requests
func (t *interceptingTransport) RoundTrip(req *http.Request) (*http.Response, error) {
	// Check if URL should be excluded from interception
	if t.shouldExclude(req.URL.String()) {
		return t.base.RoundTrip(req)
	}

	// Check if URL matches block patterns (for enforcement)
	blocked := t.isBlocked(req.URL.String())

	// Read and restore request body for logging
	var bodySnippet string
	if req.Body != nil {
		bodyBytes, err := io.ReadAll(req.Body)
		if err == nil {
			req.Body.Close()
			req.Body = io.NopCloser(bytes.NewReader(bodyBytes))

			if len(bodyBytes) > maxBodySnippet {
				bodySnippet = string(bodyBytes[:maxBodySnippet]) + "..."
			} else {
				bodySnippet = string(bodyBytes)
			}
		}
	}

	// Create event for this API call
	event := NewEvent(EventAPICall, req.Method+" "+req.URL.String()).
		WithPayload("method", req.Method).
		WithPayload("url", req.URL.String()).
		WithPayload("headers", sanitizeHeaders(req.Header)).
		WithPayload("blocked", blocked).
		WithMetadata("enforcement_mode", string(t.opts.Enforcement))

	if bodySnippet != "" {
		event = event.WithPayload("body_snippet", bodySnippet)
	}

	// Handle enforcement modes
	if blocked {
		event = event.WithPayload("enforcement_action", "blocked")

		switch t.opts.Enforcement {
		case ModeBlock:
			t.client.Track(event)
			return nil, errors.New("request blocked by Trusera policy")

		case ModeWarn:
			event = event.WithMetadata("warning", "URL matches block pattern but allowed in warn mode")
			t.client.Track(event)
			// Continue with request

		case ModeLog:
			// Just record, no action
			t.client.Track(event)
		}
	} else {
		event = event.WithPayload("enforcement_action", "allowed")
		t.client.Track(event)
	}

	// Forward request to base transport
	resp, err := t.base.RoundTrip(req)
	if err != nil {
		// Track the error
		errorEvent := NewEvent(EventAPICall, "error").
			WithPayload("method", req.Method).
			WithPayload("url", req.URL.String()).
			WithPayload("error", err.Error())
		t.client.Track(errorEvent)
		return resp, err
	}

	// Record response status
	responseEvent := NewEvent(EventAPICall, "response").
		WithPayload("method", req.Method).
		WithPayload("url", req.URL.String()).
		WithPayload("status_code", resp.StatusCode).
		WithPayload("status", resp.Status)
	t.client.Track(responseEvent)

	return resp, nil
}

// shouldExclude checks if URL matches any exclude patterns
func (t *interceptingTransport) shouldExclude(url string) bool {
	for _, pattern := range t.opts.ExcludePatterns {
		if strings.Contains(url, pattern) {
			return true
		}
	}
	return false
}

// isBlocked checks if URL matches any block patterns
func (t *interceptingTransport) isBlocked(url string) bool {
	for _, pattern := range t.opts.BlockPatterns {
		if strings.Contains(url, pattern) {
			return true
		}
	}
	return false
}

// sanitizeHeaders removes sensitive headers from logging
func sanitizeHeaders(headers http.Header) map[string]string {
	sanitized := make(map[string]string)
	sensitiveHeaders := map[string]bool{
		"authorization": true,
		"cookie":        true,
		"set-cookie":    true,
		"x-api-key":     true,
	}

	for key, values := range headers {
		lowerKey := strings.ToLower(key)
		if sensitiveHeaders[lowerKey] {
			sanitized[key] = "[REDACTED]"
		} else if len(values) > 0 {
			sanitized[key] = values[0]
		}
	}

	return sanitized
}

// CreateInterceptedClient creates a new http.Client with Trusera interception
func CreateInterceptedClient(truseraClient *Client, opts InterceptorOptions) *http.Client {
	return WrapHTTPClient(&http.Client{}, truseraClient, opts)
}

// InterceptDefault wraps http.DefaultClient with Trusera interception
func InterceptDefault(truseraClient *Client, opts InterceptorOptions) {
	http.DefaultClient = WrapHTTPClient(http.DefaultClient, truseraClient, opts)
}

// MustRegisterAndIntercept is a convenience function that registers an agent and returns an intercepted client
func MustRegisterAndIntercept(apiKey, agentName, framework string, opts InterceptorOptions) (*Client, *http.Client, error) {
	client := NewClient(apiKey)

	_, err := client.RegisterAgent(agentName, framework)
	if err != nil {
		client.Close()
		return nil, nil, fmt.Errorf("failed to register agent: %w", err)
	}

	httpClient := CreateInterceptedClient(client, opts)

	return client, httpClient, nil
}
